### PR TITLE
Cpp structures

### PR DESCRIPTION
**Pull request by gerazo**
This PR closes down the refactorings of the past 3 months on cpp exporter. The functionality is still marginal (generates skeleton mainly) but is working. Thread handling functionality was added to the runtime.

### CODE REVIEW

**Comment by devaigergely81** [comment]
First notes:
- We need to extend the examples with thread descriptions: machine, monitoring, producer_consumer, train. I'll start this in this order and commit one-by-one. You could start from the end. :)
- I suggest renaming `hu.elte.txtuml.export.cpp.description` to `hu.elte.txtuml.api.platform` analogous to `hu.elte.txtuml.api.layout`. (The C++ generation based on platform description is similar to diagram generation based on layout definition.)

File: plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/Uml2ToCppExporter.java [hunk 1]
```diff
@@ -0,0 +1,299 @@
+package hu.elte.txtuml.export.cpp;
+
+/***********************************************************
```
Comment (devaigergely81) [inline plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/Uml2ToCppExporter.java]: Please remove such comments.

**Comment by devaigergely81** [comment]
I cannot find Javadoc in the sources. At least the API (`Configuration`, `Group`) must be documented well for our users, but in the long run all Java sources need to be covered.

**Comment by gerazo** [comment]
We plan to do multiple refactorings. I have opened various issues about these. We will add docs as well.

**Comment by gerazo** [comment]
Let's also postpone the renaming as (according to the issues) we also plan to change those.

File: plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/wizardz/TxtUMLToCppWizard.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+			ThreadDescriptionExporter exporter= new ThreadDescriptionExporter();
+			exporter.exportDescription((Class<? extends Configuration>) txtUMLThreadDescription);
+			
+			if (exporter.warningListIsEmpty()) {
+				String warnings = "";
+				for (String warning : exporter.getWarnings()) {
+					warnings += warning + "\n";
+				}
+				
+				warnings += "\nWould you like continue the geeration?\n ";
```
Comment (devaigergely81) [inline plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/wizardz/TxtUMLToCppWizard.java]: Misspelled: "geeration"

**Comment by devaigergely81** [comment]
Simple configuration for the Machine1 example:
~~~~
@Group(contains = {Machine.class, User.class})
@Multithreading(false)
public class Machine1Platform extends Configuration {}
~~~~
Compilation fails with this:
~~~~
java.lang.NullPointerException
  java.lang.NullPointerException
  	at org.eclipse.core.internal.runtime.Activator.getURLConverter(Activator.java:324)
  	at org.eclipse.core.runtime.FileLocator.toFileURL(FileLocator.java:206)
  	at hu.elte.txtuml.export.cpp.Uml2ToCppExporter.seekCppFilesColcations(Uml2ToCppExporter.java:183)
  	at hu.elte.txtuml.export.cpp.Uml2ToCppExporter.copyPreWrittenCppFiles(Uml2ToCppExporter.java:138)
  	at hu.elte.txtuml.export.cpp.Uml2ToCppExporter.buildCppCode(Uml2ToCppExporter.java:111)
  	at hu.elte.txtuml.export.cpp.wizardz.TxtUMLToCppWizard.performFinish(TxtUMLToCppWizard.java:100)
  	at org.eclipse.jface.wizard.WizardDialog.finishPressed(WizardDialog.java:799)
  	at org.eclipse.jface.wizard.WizardDialog.buttonPressed(WizardDialog.java:429)
  	at org.eclipse.jface.dialogs.Dialog$2.widgetSelected(Dialog.java:619)
  	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:248)
  	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
  	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4362)
  	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1113)
  	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4180)
  	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3769)
  	at org.eclipse.jface.window.Window.runEventLoop(Window.java:827)
  	at org.eclipse.jface.window.Window.open(Window.java:803)
  	at hu.elte.txtuml.export.cpp.handlers.TxtUMLToCppHandler.execute(TxtUMLToCppHandler.java:19)
  	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:295)
  	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:90)
  	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
  	at java.lang.reflect.Method.invoke(Unknown Source)
  	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:56)
  	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:252)
  	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:234)
  	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:132)
  	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:152)
  	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:493)
  	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:486)
  	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:210)
  	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.executeItem(HandledContributionItem.java:799)
  	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.handleWidgetSelection(HandledContributionItem.java:675)
  	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.access$7(HandledContributionItem.java:659)
  	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem$4.handleEvent(HandledContributionItem.java:592)
  	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
  	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4362)
  	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1113)
  	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4180)
  	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3769)
  	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1127)
  	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:337)
  	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1018)
  	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:156)
  	at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:654)
  	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:337)
  	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:598)
  	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
  	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:139)
  	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
  	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
  	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
  	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:380)
  	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:235)
  	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
  	at java.lang.reflect.Method.invoke(Unknown Source)
  	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:669)
  	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:608)
  	at org.eclipse.equinox.launcher.Main.run(Main.java:1515)
  	at org.eclipse.equinox.launcher.Main.main(Main.java:1488)
~~~~~

**Comment by devaigergely81** [comment]
I propose not to postpone the refactorings that are visible for the user: API documentation, API naming. The master branch must remain in releasable state, but this branch is currently far from being releasable, according to my first impressions.
Please review any UI messages: Almost all messages I got so far contained typos.

File: plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/wizardz/TxtUMLToCppWizard.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+			}
+				
+			Uml2ToCppExporter cppExporter = new Uml2ToCppExporter(model,exporter.getConfigMap(),exporter.isMultiThreading(),runtimeOption,debugOption);
+			try {
+				cppExporter.buildCppCode(projectFolder + File.separator + 
+						GenericFolderName + File.separator + txtUMLModel);
+				
+				
+			} catch (Exception e) {
+				Dialogs.errorMsgb("Completion faild", e.getMessage(), e);
```
Comment (devaigergely81) [inline plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/wizardz/TxtUMLToCppWizard.java]: Sould be: "Compilation failed"

File: plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/wizardz/TxtUMLToCppWizard.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+			}
+				
+			Uml2ToCppExporter cppExporter = new Uml2ToCppExporter(model,exporter.getConfigMap(),exporter.isMultiThreading(),runtimeOption,debugOption);
+			try {
+				cppExporter.buildCppCode(projectFolder + File.separator + 
+						GenericFolderName + File.separator + txtUMLModel);
+				
+				
+			} catch (Exception e) {
+				Dialogs.errorMsgb("Completion failed", e.getMessage(), e);
```
Comment (devaigergely81) [inline plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/wizardz/TxtUMLToCppWizard.java]: Shouldn't this be "Compilation failed"?

**Comment by nagyandras95** [comment]
The Machine1 example is working for me by the same configuration you created but I have used guest eclipse to try it. As I see the problem caused by seeking the pre-written cpp files. I guess the method tries to search them on the file-system and does not handle like a plug-in resource. Maybe they could be downloadable so they is not needed to be generated for each project.

**Pull request closed without merging**